File: csharp/exercise/exer07-exec-service-v0a.cs
/*
 * MY EXECUTOR SERVICE
 *
 * Version 0A: The easiest executor service
 * - It uses a blocking queue as underlying mechanism.
 */
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;



namespace Exer07
{
    class MyExecServiceV0A
    {
        private int numThreads = 0;
        private List<Thread> lstTh = new List<Thread>();
        private BlockingCollection<IRunnable> taskPending = new BlockingCollection<IRunnable>();



        public MyExecServiceV0A(int numThreads) {
            init(numThreads);
        }



        private void init(int inpNumThreads)
        {
            numThreads = inpNumThreads;

            for (int i = 0; i < numThreads; ++i)
                lstTh.Add(new Thread(() => threadWorkerFunc(this)));

            lstTh.ForEach(th => th.Start());
        }



        public void submit(IRunnable task)
        {
            taskPending.Add(task);
        }



        public void waitTaskDone()
        {
            // This ExecService is too simple,
            // so there is no implementation for waitTaskDone()
            Thread.Sleep(11000); // fake behaviour
        }



        public void shutdown()
        {
            // This ExecService is too simple,
            // so there is no implementation for shutdown()
            Console.WriteLine("No implementation for shutdown().");
            Console.WriteLine("You need to exit the app manually.");
        }



        private static void threadWorkerFunc(MyExecServiceV0A thisPtr)
        {
            ref var taskPending = ref thisPtr.taskPending;
            IRunnable task = null;

            for (; ; )
            {
                // WAIT FOR AN AVAILABLE PENDING TASK
                task = taskPending.Take();

                // DO THE TASK
                task.run();
            }
        }
    }
}

File: python/exer07_exec_service_v0a.py
'''
MY EXECUTOR SERVICE

Version 0A: The easiest executor service
- It uses a blocking queue as underlying mechanism.
'''

import time
from queue import Queue
import threading
from exer07_exec_service_itask import ITask



class MyExecServiceV0A:
    class EmptyTask(ITask):
        def run(self):
            pass


    def __init__(self, num_threads: int):
        self.__num_threads = num_threads
        self.__lstth = []
        self.__task_pending = Queue()

        for _ in range(self.__num_threads):
            self.__lstth.append(
                threading.Thread(target=MyExecServiceV0A.__thread_worker_func, args=(self,))
            )

        for th in self.__lstth:
            th.start()


    def submit(self, task: ITask):
        self.__task_pending.put_nowait(task)


    def wait_task_done(self):
        # This ExecService is too simple,
        # so there is no implementation for waitTaskDone()
        time.sleep(11) # fake behaviour


    def shutdown(self):
        # This ExecService is too simple,
        # so there is no implementation for shutdown()
        print('No implementation for shutdown().')
        print('You need to exit the app manually.')


    @staticmethod
    def __thread_worker_func(selfptr: 'MyExecServiceV0A'):
        task_pending = selfptr.__task_pending

        while True:
            # WAIT FOR AN AVAILABLE PENDING TASK
            task = task_pending.get()

            # DO THE TASK
            task.run()
